R/test-threads.R

```diff
@@ -0,0 +1,141 @@
+#' Takes a model object and runs it with various threads values
+#'
+#' @param .mod bbi_model object to copy/test.
+#' @param threads Integer vector of threads values to test.
+#' @param .mode Passed through to bbr::submit_models(.mode).
+#' @param .bbi_args a named list.
+#'
+#' @importFrom checkmate assert_list
+#'
+#' @return A list of the model objects for the submitted models.
+#'
+#' @export
+test_threads <- function(
+  .mod,
+  threads = c(2,4),
+  .mode = getOption("bbr.bbi_exe_mode"),
+  .bbi_args = list()
+) {
+
+  check_model_object(.mod)
+  assert_list(.bbi_args)
+
+  .mods <- map(threads, ~ copy_model_from(.mod, paste0(get_model_id(.mod), "_", .x, "_threads")) %>%
+                 add_bbi_args(.bbi_args = c(threads = .x,
+                                            .bbi_args,
+                                            parallel = TRUE,
+                                            overwrite = TRUE)) %>%
+                 add_tags(paste("test",.x,"threads")))
+
+  mod_paths <- lapply(.mods, function(mod.x){mod.x$absolute_model_path}) %>% unlist()
+
+  submit_models(.mods, .mode = .mode, .wait = FALSE)
+
+  .mods
+}
+
+#' Check estimation time for models run with various threads values
+#'
+#' @param .mods list of bbi model objects created by `test_threads()`.
+#' @param return_times character vector denoting which times from `model_summary()` you want to return.
+#'        See details for more information.
+#' @param wait logical. If TRUE, pass `.mods` to `wait_for_nonmem()` before returning results.
+#' @param ... args passed through to `wait_for_nonmem()`.
+#'
+#' @details
+#' `return_times` can be any subset of `c("estimation_time", "covariance_time", "cpu_time")`.
+#' Users can also specify `all`, which is the shorthand method for selecting all 3 of those columns
+#'
+#' @examples
+#' \dontrun{
+#' mods <- test_threads(mod, threads = c(2, 4))
+#'
+#' If models have not finished:
+#' check_threads(mods, wait = TRUE, time_limit = 300)
+#' check_threads(mods, wait = TRUE, return_times = c("estimation_time", "covariance_time"))
+#'
+#' If models have already finished:
+#' check_threads(mods, wait = FALSE)
+#' check_threads(mods, wait = FALSE, return_times = "All")
+#' }
+#' @return A tibble with columns `threads` (number of threads) and `time`
+#'   (elapsed estimation time in seconds for test models).
+#'
+#' @importFrom tidyselect all_of
+#'
+#' @export
+check_threads <- function(
+  .mods,
+  return_times = "estimation_time",
+  wait = TRUE,
+  ...
```

Comment:
> Im also curious if you think the default for wait in check_threads should be TRUE or not

Yes, I think so. We want people to be able to do

```r
mod %>%
  test_threads() %>%
  check_threads()
```
and have it likely work.

Reply:
Yup had the same thought